Build/QIR_validation.ps1

```diff
@@ -0,0 +1,118 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+$ErrorActionPreference = 'Stop'
+
+& "$PSScriptRoot/set-env.ps1"
+$all_ok = $True
+
+function Build-One {
+    param(
+        $project
+    );
+
+    $project_directory = $(Split-Path $project -Parent)
+    $project_name = $(Split-Path $project -LeafBase)
+
+    pushd $project_directory
+
+    Write-Host "##[info]Building $project..."
+    dotnet build $project `
+        -c $Env:BUILD_CONFIGURATION `
+        -v $Env:BUILD_VERBOSITY `
+        --no-dependencies `
+        /property:DefineConstants=$Env:ASSEMBLY_CONSTANTS `
+        /property:Version=$Env:ASSEMBLY_VERSION `
+        /property:QirGeneration="true"
+
+    if  ($LastExitCode -ne 0) {
+        Write-Host "##vso[task.logissue type=error;]Failed to build $project."
+        $script:all_ok = $False
+    } else {
+        qir-cli `
+            --dll (Join-Path bin $Env:BUILD_CONFIGURATION netcoreapp3.1 "${project_name}.dll") `
+            --exe qir
+        if  ($LastExitCode -ne 0) {
+            Write-Host "##vso[task.logissue type=error;]Failed to build $project."
+            $script:all_ok = $False
+        } else {
+            Get-ChildItem (Join-Path qir *__Interop.exe) `
+                | ForEach-Object { & $_ }
+            if  ($LastExitCode -ne 0) {
+                Write-Host "##vso[task.logissue type=error;]$project encountered an error or failed during execution."
+                $script:all_ok = $False
+            }
+        }
+    }
+
+    if ($env:FORCE_CLEANUP -eq "true") {
+        # Force cleanup of generated bin, obj, and qir folders for this project.
+        Write-Host "##[info]Cleaning up bin/obj from $project_directory..."
+        Get-ChildItem -Path $project_directory -Recurse | Where-Object { ($_.name -eq "bin" -or $_.name -eq "obj" -or $_.name -eq "qir") -and $_.attributes -eq "Directory" } | Remove-Item -recurse -force
+    }
+
+    popd
+}
+
+# The commented out lines are sample projects that are not yet compatible for QIR generation/execution. 
```

Comment:
It would be great to figure out how to run samples that need arguments in a programmatic way. I guess the alternative would be to hardcode the needed arguments for each sample that needs them. If that is the direction to go it, I would need some help with the powershell in learning how to do that.

Reply:
If by hardcoding you mean hardcoding the values in this powershell script, then I think that makes sense.

Reply:
One way to do that is to store a PowerShell array of needed arguments with each project listed here, then to splat those arguments into the call above.

https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_splatting

Reply:
I'll try to be more specific here. What I am looking for is some way to zip the arguments and their values together with the path to the csproj file in the big list of samples. So instead of having
```
$QirProjects = {
    <path to sample 1>,
    <path to sample 2>,
    ...
}
```
I would have something like
```
$QirProjects = {
    { Path = <path to sample 1>, Args = [(arg1,val1), (arg2,val2), ...] },
    { Path = <path to sample 2>, Args = [(arg1,val1), (arg2,val2), ...] },
    ...
}
```
but I'm not seeing a straightforward way of accomplishing that. I expect that something like this is built in to powershell, or that maybe I'm just not thinking about the problem in a powershell-like way.

Reply:
I guess that might be further complicated by the fact that any given sample may have multiple entry points, resulting in multiple executables to be called, possibly with different arguments.

Reply:
Setting aside the multiple entry points part, your snippet above is almost what you need:

```powershell
$qirProjects = @(
    @{ "Path" = "path to sample 1"; "Args" = @("--arg-1", "val1", "--arg-2", "val2"); }
    ...
);
```

You can then splat in `$_["Args"]` in your `ForEach-Object` script blocks, using the splatting syntax linked to above.

Reply:
Oh a hash table, yah that's perfect, exactly what I need. Thanks @cgranade! I can see now what is meant by 'splat'. It is similar to the star syntax from python or 'explode' from other languages.